# KNN From Scratch vs Scikit-learn

In [271]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
from sklearn.neighbors import KNeighborsClassifier

In [272]:
df = pd.read_csv("Social_Network_Ads 2.csv")

In [273]:
df.head()

,User ID,Gender,Age,EstimatedSalary,Purchased
0,15624510,Male,19,19000,0
1,15810944,Male,35,20000,0
2,15668575,Female,26,43000,0
3,15603246,Female,27,57000,0
4,15804002,Male,19,76000,0


## Dataset

- Social Network Ads dataset  
- Features:
  - Gender (encoded)
  - Age
  - Estimated Salary  
- Target:
  - Purchased (0 or 1)


In [274]:
df.drop(columns='User ID',inplace=True)
df.head()

,Gender,Age,EstimatedSalary,Purchased
0,Male,19,19000,0
1,Male,35,20000,0
2,Female,26,43000,0
3,Female,27,57000,0
4,Male,19,76000,0


In [275]:
from sklearn.preprocessing import LabelEncoder , StandardScaler
encoder= LabelEncoder()
df['Gender']= encoder.fit_transform(df['Gender'])
print(df)

     Gender  Age  EstimatedSalary  Purchased
0         1   19            19000          0
1         1   35            20000          0
2         0   26            43000          0
3         0   27            57000          0
4         1   19            76000          0
..      ...  ...              ...        ...
395       0   46            41000          1
396       1   51            23000          1
397       0   50            20000          1
398       1   36            33000          0
399       0   49            36000          1

[400 rows x 4 columns]


In [276]:
scaler= StandardScaler()
X=df.iloc[:,0:3]
y=df.iloc[:,-1].values

In [277]:
X=scaler.fit_transform(X)
X

array([[ 1.02020406, -1.78179743, -1.49004624],
       [ 1.02020406, -0.25358736, -1.46068138],
       [-0.98019606, -1.11320552, -0.78528968],
       ...,
       [-0.98019606,  1.17910958, -1.46068138],
       [ 1.02020406, -0.15807423, -1.07893824],
       [-0.98019606,  1.08359645, -0.99084367]])

In [278]:
y

array([0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0,
       1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0,
       1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1,

In [279]:
from sklearn.model_selection import train_test_split 
X_train,X_test,y_train,y_test=train_test_split(X,y,random_state=0,test_size=0.2)

In [280]:
knn= KNeighborsClassifier(n_neighbors=5)

In [281]:
knn.fit(X_train,y_train)

KNeighborsClassifier()

In [282]:
y_pred=knn.predict(X_test)

In [283]:
from sklearn.metrics import accuracy_score, r2_score

In [284]:
print (accuracy_score(y_test,y_pred))
print (r2_score(y_test,y_pred))

0.95
0.7492163009404389


In [285]:
X_test.shape

(80, 3)

In [286]:
X_train.shape

(320, 3)

## Distance Metric

Euclidean Distance:

d(x, y) = √Σ(xᵢ - yᵢ)²

- Used to measure similarity between points  
- Smaller distance → more similar  

In [287]:
from collections import Counter
class KNN():
    def __init__(self,k=5):
        self.k=k
        self.X_train=None 
        self.y_train=None
    def fit(self,X_train,y_train):
        self.X_train= X_train
        self.y_train= y_train
    def predict(self,X_test):
        y_pred=[]
        for i in X_test:
            distances=[]
            for j in self.X_train:
                distances.append(self.calculate_distance(i,j))
            n_neighbours=sorted(list(enumerate(distances)),key=lambda x:x[1])[0:self.k]
            label=self.majority_count(n_neighbours)
            y_pred.append(label)
        return np.array(y_pred)

    def calculate_distance(self,point_A,point_B):
        return np.linalg.norm(point_A-point_B)

    def majority_count(self,neighbours):
        votes=[]
        for i in neighbours : 
            votes.append(self.y_train[i[0]])
        votes= Counter(votes)
        return votes.most_common()[0][0]
    


In [288]:
myknn= KNN()
myknn.fit(X_train,y_train)

In [289]:
y_pred1=myknn.predict(X_test)

In [290]:
accuracy_score(y_test,y_pred1)

0.95

In [291]:
r2_score(y_test,y_pred1)

0.7492163009404389

## Results

| Model              | Accuracy | R² Score |
|--------------------|----------|----------|
| Custom KNN         | 0.95     | 0.749    |
| Scikit-learn KNN   | 0.95     | 0.749    |

## Observations

- Both models give **identical results**
- Confirms correctness of custom implementation
- Predictions match exactly
